File: 2025/1.py
MAX = 100


def modulo_minus(dial, distance, part2: bool = False):
    num_zeros = 0

    def minus(dial, distance, zeros):
        was_positive = dial > 0

        dial -= distance

        if dial == 0:
            zeros += 1

        if dial < 0:
            dial += MAX
            if was_positive and part2:
                zeros += 1

        return dial, zeros

    if abs(distance) > MAX and part2:
        num_zeros += distance // MAX

    dial, num_zeros = minus(dial, distance % MAX, num_zeros)
    return dial, num_zeros


def modulo_plus(dial, distance, part2: bool = False):
    num_zeros = 0

    def plus(dial, distance, zeros):
        dial += distance

        if dial == MAX:
            dial = 0
            zeros += 1

        if dial > MAX:
            dial -= MAX
            if part2:
                zeros += 1

        return dial, zeros

    if abs(distance) > MAX and part2:
        num_zeros += distance // MAX

    dial, num_zeros = plus(dial, distance % MAX, num_zeros)
    return dial, num_zeros


def test():
    inputs = [
        "L68",
        "L30",
        "R48",
        "L5",
        "R60",
        "L55",
        "L1",
        "L99",
        "R14",
        "L82",
    ]

    dial = 50
    num_zeros = 0

    for line in inputs:
        distance = int(line[1:])

        if line[0] == "L":
            t = modulo_minus(dial, distance, True)
            dial = t[0]
            num_zeros += t[1]

        elif line[0] == "R":
            t = modulo_plus(dial, distance, True)
            dial = t[0]
            num_zeros += t[1]

        print(f"After {line}, dial is at {dial} and {num_zeros=}")

    print(num_zeros)


def part1(lines: list[str]):
    dial = 50
    num_zeros = 0

    for line in lines:
        distance = int(line[1:])

        if line[0] == "L":
            t = modulo_minus(dial, distance)
            dial = t[0]
            num_zeros += t[1]

        elif line[0] == "R":
            t = modulo_plus(dial, distance)
            dial = t[0]
            num_zeros += t[1]

        # print(f"After {line[0]}{int(line[1:])}, dial is {dial}")

    print(num_zeros)


def part2(lines: list[str]):
    dial = 50
    num_zeros = 0

    for line in lines:
        distance = int(line[1:])

        if line[0] == "L":
            t = modulo_minus(dial, distance, part2=True)
            dial = t[0]
            num_zeros += t[1]

        elif line[0] == "R":
            t = modulo_plus(dial, distance, part2=True)
            dial = t[0]
            num_zeros += t[1]

        # print(f"After {line[0]}{int(line[1:])}, dial is {dial}")

    print(num_zeros)

def main():
    with open("day1.in") as f:
        lines = f.readlines()

    part1(lines)
    part2(lines)


if __name__ == "__main__":
    # test()
    main()

File: 2025/1.rs
use std::fs;

fn main() {
    let input: Vec<String> = fs::read_to_string("day1.in")
        .unwrap()
        .lines()
        .map(String::from)
        .collect();

    part1(input.clone());
    // part2(input.clone());
}

fn part1(input: Vec<String>) {
    let mut dial: i32 = 50;
    let mut num_zeros: i32 = 0;

    for line in input {
        let rotation = &line[0..1];
        let distance: i32 = match line[1..line.len()].parse() {
            Ok(num) => num,
            Err(e) => {
                eprintln!("Failed to parse integer: {}", e);
                // Handle the error as appropriate for your application
                return;
            }
        };

        if rotation == "L" {
            dial = (dial - distance) % 100;
        } else if rotation == "R" {
            dial = (dial + distance) % 100;
        }

        println!("After {}{}, dial is {}", rotation, distance, dial);

        if dial == 0 {
            num_zeros += 1;
        }
    }

    println!("Part 1 password is: {}", num_zeros);
}

fn modulo_minus(mut dial: i32, distance: i32) -> (i32, i32) {
    let mut num_zeros = 0;

    // Set the dial to maximum value if it is at 0
    if dial == 0 {
        dial = 100;
    }

    if (dial - distance) >= 0 {
        dial -= distance;
        if dial == 0 {
            num_zeros += 1;
        }
    } else {
        // Case when dial < distance
        dial -= distance;
        if dial < 0 {
            num_zeros += 1;
            // Apply the modulo operation
            dial += 100;
        }
    }

    return (dial, num_zeros);
}

fn part2(input: Vec<String>) {
    let mut dial: i32 = 50;
    let mut num_zeros: i32 = 0;

    for line in input {
        let rotation = &line[0..1];
        let distance: i32 = match line[1..line.len()].parse() {
            Ok(num) => num,
            Err(e) => {
                eprintln!("Failed to parse integer: {}", e);
                // Handle the error as appropriate for your application
                return;
            }
        };

        if rotation == "L" {
            if distance < 100 {
                let t = modulo_minus(dial, distance);
                dial = t.0;
                num_zeros = t.1;
            } else {
                for _ in 0..(distance / 100) {
                    let d: i32 = 100;
                    let t = modulo_minus(dial, d);
                    dial = t.0;
                    num_zeros = t.1;
                }
            }
        } else if rotation == "R" {
            dial = (dial + distance) % 100;
        }

        if dial == 0 {
            num_zeros += 1;
        }
    }

    println!("Part 2 password is: {}", num_zeros);
}
